File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/InventoryQuery/Filters/StarRatingRangeFilterDecorator.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.InventoryQuery.Filters;

/// <summary>
///     A filter that filters inventory items by a range of star ratings.
/// </summary>
public sealed class StarRatingRangeFilterDecorator : InventoryQueryDecoratorBase
{
    private readonly int? _max;
    private readonly int? _min;

    /// <summary>
    ///     Initializes a new instance of the <see cref="StarRatingRangeFilterDecorator" /> class.
    /// </summary>
    /// <param name="inner">The inner query to decorate.</param>
    /// <param name="minInclusive">The minimum star rating (inclusive), or null for no minimum.</param>
    /// <param name="maxInclusive">The maximum star rating (inclusive), or null for no maximum.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown when ratings are not between 1 and 5.</exception>
    /// <exception cref="ArgumentException">Thrown when min is greater than max.</exception>
    public StarRatingRangeFilterDecorator(IInventoryQuery inner, int? minInclusive, int? maxInclusive) : base(inner)
    {
        if (minInclusive is not null && (minInclusive < 1 || minInclusive > 5))
            throw new ArgumentOutOfRangeException(nameof(minInclusive), "Rating must be between 1 and 5");
        if (maxInclusive is not null && (maxInclusive < 1 || maxInclusive > 5))
            throw new ArgumentOutOfRangeException(nameof(maxInclusive), "Rating must be between 1 and 5");
        if (minInclusive is not null && maxInclusive is not null && minInclusive > maxInclusive)
            throw new ArgumentException("minInclusive cannot be greater than maxInclusive.");

        _min = minInclusive;
        _max = maxInclusive;
    }

    /// <inheritdoc />
    protected override string? AppliedDescription
    {
        get
        {
            if (_min is null && _max is null) return null;
            if (_min is not null && _max is not null)
                return $"Filter: Star rating between {_min.Value} and {_max.Value} (inclusive)";
            if (_min is not null)
                return $"Filter: Star rating >= {_min.Value}";
            return $"Filter: Star rating <= {_max!.Value}";
        }
    }

    /// <inheritdoc />
    protected override IReadOnlyList<InventoryItem> Decorate(IReadOnlyList<InventoryItem> items)
    {
        IEnumerable<InventoryItem> filtered = items;

        if (_min is not null) filtered = filtered.Where(i => i.StarRating.Rating >= _min.Value);
        if (_max is not null) filtered = filtered.Where(i => i.StarRating.Rating <= _max.Value);

        return filtered.ToList().AsReadOnly();
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventoryquery/filters/StarRatingRangeFilterDecorator.java
package assignment2solution.domain.inventoryquery.filters;

import assignment2solution.domain.inventory.InventoryItem;
import assignment2solution.domain.inventoryquery.IInventoryQuery;
import assignment2solution.domain.inventoryquery.InventoryQueryDecoratorBase;

import java.util.List;
import java.util.stream.Collectors;

/**
 * A filter that filters inventory items by a range of star ratings.
 */
public final class StarRatingRangeFilterDecorator extends InventoryQueryDecoratorBase {
    private final Integer min;
    private final Integer max;

    /**
     * Initializes a new instance of the {@link StarRatingRangeFilterDecorator} class.
     */
    public StarRatingRangeFilterDecorator(IInventoryQuery inner, Integer minInclusive, Integer maxInclusive) {
        super(inner);
        if (minInclusive != null && (minInclusive < 1 || minInclusive > 5)) {
            throw new IllegalArgumentException("Rating must be between 1 and 5");
        }
        if (maxInclusive != null && (maxInclusive < 1 || maxInclusive > 5)) {
            throw new IllegalArgumentException("Rating must be between 1 and 5");
        }
        if (minInclusive != null && maxInclusive != null && minInclusive > maxInclusive) {
            throw new IllegalArgumentException("minInclusive cannot be greater than maxInclusive.");
        }
        this.min = minInclusive;
        this.max = maxInclusive;
    }

    @Override
    protected String getAppliedDescription() {
        if (min == null && max == null) {
            return null;
        }
        if (min != null && max != null) {
            return "Filter: Star rating between " + min + " and " + max + " (inclusive)";
        }
        if (min != null) {
            return "Filter: Star rating >= " + min;
        }
        return "Filter: Star rating <= " + max;
    }

    @Override
    protected List<InventoryItem> decorate(List<InventoryItem> items) {
        var stream = items.stream();

        if (min != null) {
            stream = stream.filter(i -> i.getStarRating().getRating() >= min);
        }
        if (max != null) {
            stream = stream.filter(i -> i.getStarRating().getRating() <= max);
        }

        return stream.collect(Collectors.toUnmodifiableList());
    }
}
